controllers/workplacement_controller.go

```diff
@@ -120,15 +119,20 @@ func (r *WorkPlacementReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	if versionID != "" && workPlacement.Status.VersionID != versionID {
 		workPlacement.Status.VersionID = versionID
+		logger.Info("Updating version status", "versionID", versionID)
 		err = r.Client.Status().Update(ctx, workPlacement)
 		if err != nil {
 			r.VersionCache[workPlacement.GetUniqueID()] = versionID
 			logger.Error(err, "Error updating WorkPlacement status")
 			return ctrl.Result{}, err
 		}
 	}
-	logger.Info("WorkPlacement successfully reconciled", "workPlacement", workPlacement.Name, "versionID", versionID)
 
+	if missingFinalizers := checkWorkPlacementFinalizers(workPlacement, filepathMode); len(missingFinalizers) > 0 {
+		return addFinalizers(opts, workPlacement, missingFinalizers)
+	}
+
+	logger.Info("WorkPlacement successfully reconciled", "workPlacement", workPlacement.Name, "versionID", versionID)
```

Comment:
the order is now:

1. update the files in statestore
1. update the status of the workplacement with the latest VersionID
1. update the finalizers

I considered swapping 2/3 round, i dont mind tbh. They both can't be done at the same time annoyingly since its an update to the status and the metadata, which is two separate API calls, which won't work without requeuing since the resource will have changed between the two API calls